File: Coursework/Numerical Analysis/Methods_First_Exam/forward_euler.jl
using LinearAlgebra

function forward_euler(A, y0, t, h)
    """
    Solve y' = Ay with y(0) = y0 by forward Euler method
    
    Args:
        A: matrix
        y0: initial condition for y
        t: 1D array of t values where we approximate y values. Time step
            at each iteration is given by t[n+1] - t[n].
        h: time step
        
    Returns:
        y: 2D array
            Approximation of the solution y(t) computed at each time in t.
    """
    
    y = zeros(eltype(y0), length(y0), length(t))
    y[:,1] = y0
    for i in 2:length(t)
        y[:,i] = y[:,i-1] + h * A * y[:,i-1]
    end
    return y
end

File: Coursework/Numerical Analysis/Midterm-1/forward_euler.py
import numpy as np

def forward_euler(A, y0, t, h):
    """
    Solve y' = Ay with y(0) = y0 by forward Euler method

    Args:
        A: matrix
        y0: initial condition for y
        t: 1D NumPy array of t values where we approximate y values. Time step
            at each iteration is given by t[n+1] - t[n].
        h: time step
    
    Returns:
        y: 2D NumPy array
            Approximation of the solution y(t) computed at each time in t.
    """
    
    y = np.zeros((len(y0), len(t)))
    y[:,0] = y0
    for i in range(1, len(t)):
        y[:,i] = y[:,i-1] + h*A@y[:,i-1]
    return y